### ISSUE TITLE
Preview mode works differently under the Django 1.10 middleware system

### ISSUE THREAD

**Issue by coredumperror** [wagtail/wagtail#3850]
This is very much the same bug as #3546, but I want to bring it up separately because I'm having a completely different problem. Every time one of my users previews a page with an image on it, the preview crashes.

The problem, as @gasman mentioned in #3546, is that Wagtail can no longer do the same thing it did with Django 1.9-style middleware. It would run the `process_request()` method for each middleware, but not `process_response()`. Under the Django 1.10 middleware mechanism, it cannot skip the `process_response()` calls, and that's what's breaking my site.

My site uses [CrequestMiddleware](https://github.com/Alir3z4/django-crequest) to store a globally accessible `request` object for the duration of each request. We use that request for a variety of purposes, including figuring out what path uploaded files should be saved to in our S3 bucket. We need to differentiate between files uploaded to different Sites, so that Site A can upload `file.jpg` without interfering with Site B's `file.jpg` that's already there.

For what I assume is necessary anti-memory-leak reasons, CrequestMiddleware destroys the request object during `process_response()`. This is a huge problem because it stores said request object in a dictionary keyed by the current thread object (the return value of `threading.current_thread()`). Since `Page.dummy_request()` invokes `BaseHandler.middleware_chain()` while Django's original call to middleware_chain() is still under way, this causes the stored request object to be overwritten (in the thread's second call to `CrequestMiddleware.process_request()` and then destroyed (in the thread's first call to `CrequestMiddleware.process_response()` before the preview is rendered.

This makes it impossible to retrieve the stored request object when rendering the image derivative, which throws an exception. Then the original middleware chain unwinds, and `CrequestMiddleware.process_response()` is called again, but it has nothing to delete, because the stored request is already gone.

The only solution I can think of for this would be to have Wagtail mark the dummy request object with some sort of flag (like it used to do with the `request.is_preview` attribute). I could then subclass `CrequestMiddleware` to alter `process_response()` to not delete the stored request when `request.is_preview == True`.

That feels pretty hacky, though. Especially since it makes a third-party middleware need to be aware of a Wagtail-specific flag. I'd love for there to be a better solution, but nothing is coming to mind. 

I'll go whip up a PR for that `is_preview` flag.

**Comment by coredumperror** [comment]
I ended up going for `request.is_dummy` instead of `is_preview`, because `is_preview` is already in use in `Page.serve_preview` and fulfills a slightly different purpose.

**Comment by coredumperror** [comment]
After discussing the problem with my colleagues, I've found an even better solution, and it might even be good enough to incorporate into wagtail itself as a workaround for this problem.

Rather than altering `Page.dummy_request()`, we add a new middleware at the top of the `MIDDLEWARE` list, and subclass problematic middleware like so:

```python
class MiddlewareIterationCounter(object):
    """
    This middleware exists to prevent problems with Wagtail Page previews. Page.dummy_request()
    re-executes the entire middleware chain, which middleware isn't written to expect. For example,
    CrequestMiddleware only expects to ever run process_response() once per thread, which is why it
    assumes it's safe to delete the stored request when it runs. But since the middleware chain gets
    run twice during a preview, it deletes the current thread's stored request before the request 
    cycle is actually finished.

    By counting the interations of the middleware chain, we can subclass problematic middleware to 
    make them exclude themselves from subsequent iterations.
    """

    _iterations = defaultdict(int)

    def __init__(self, get_response=None):
        self.get_response = get_response

    def __call__(self, request):
        # Increment the current thread's interation counter on ingress.
        self._iterations[threading.current_thread()] += 1
        response = self.get_response(request)
        # Decrement the current thread's interation counter on egress.
        self._iterations[threading.current_thread()] -= 1
        return response

    @classmethod
    def get_iteration_count(cls):
        """
        Returns the iteration count for the current thread.
        Since cls._iterations is a defaultdict, this can safely be called before the first increment, which returns 0.
        """
        return cls._iterations[threading.current_thread()]


class CustomCrequestMiddleware(CrequestMiddleware):

    def __init__(self, get_response=None):
        # Skip this middleware if the middleware chain has already run at least once.
        if MiddlewareIterationCounter.get_iteration_count() >= 1:
            raise MiddlewareNotUsed
        super(CustomCrequestMiddleware, self).__init__(get_response)
```

Adding `MiddlewareIterationCounter` to your `MIDDLEWARE` list makes it count the number of active iterations of the middelware chain within a single thread, decrementing the counter when the chain finishes processing the response. By subclassing `CrequestMiddleware` to pay attention to that counter, we can prevent it from inserting itself into the middleware chain that `dummy_request()` builds and executes.

The position in which you place `MiddlewareIterationCounter` into `MIDDLEWARE` doesn't actually matter, because all of the middleware get constructed (in `BaseHandler.load_middleware()`) before the middleware chain starts. So as long as `MiddlewareIterationCounter` is anywhere in the list, it will increment its counter to `1` on ingress before `dummy_request()` has a change to run `load_middleware()`, which is when the check in `CustomCrequestMiddleware.__init__()` triggers and excludes it from the second chain.

**Comment by coredumperror** [comment]
I'm curious why you think `CrequestMiddleware`'s behavior is at fault here. Is there a better way for it to avoid memory a leak, besides deleting the stored request object during `process_response()`?

**Comment by gasman** [comment]
To be honest, I disagree with the whole premise of CrequestMiddleware - that the request object should be available everywhere - so I can't really comment on better ways to achieve that! Model code shouldn't be aware of the environment it's being called from - it should work the same whether it's being called from a web request, a management command, a task queue or anywhere else. If a model needs to behave differently depending on the site it belongs to, then the concept of 'belonging to a site' should be made explicit in the model (possibly as a 'site' property, which may or may not correspond to an actual database field) and the view code should actively pass it the relevant site, rather than the model trying to pick it up from the environment.

I realise things often aren't that perfect in the real world, and if you're working with the Wagtail admin the views might just not be set up that way, in which case CrequestMiddleware might be a necessary evil. But ultimately it's still trying to fight against a deliberate design feature of Django, which means it inevitably has to make some not-entirely-reliable assumptions such as "each thread will only be dealing with one request object at a time".

**Comment by coredumperror** [comment]
Fair enough. I think it's a cripplingly dumb design decision, but I agree that working around it does imply that sometimes things will go wrong.
